xero/auth.py

```diff
@@ -6,20 +6,22 @@
 from oauthlib.oauth1 import (
     SIGNATURE_RSA, SIGNATURE_TYPE_AUTH_HEADER, SIGNATURE_HMAC
 )
-from requests_oauthlib import OAuth1
+from requests_oauthlib import OAuth1, OAuth2Session, OAuth2
 from six.moves.urllib.parse import urlencode, parse_qs
 
 from .constants import (
-    XERO_BASE_URL, REQUEST_TOKEN_URL, AUTHORIZE_URL, ACCESS_TOKEN_URL
+    XERO_BASE_URL, REQUEST_TOKEN_URL, AUTHORIZE_URL, ACCESS_TOKEN_URL,
+    OAUTH2_AUTHORIZE_URL, OAUTH2_TOKEN_URL, OAUTH2_CONNECTIONS_URL
 )
 from .exceptions import (
     XeroBadRequest, XeroException, XeroExceptionUnknown, XeroForbidden,
     XeroInternalError, XeroNotAvailable, XeroNotFound, XeroNotImplemented,
-    XeroNotVerified, XeroRateLimitExceeded, XeroUnauthorized
+    XeroNotVerified, XeroRateLimitExceeded, XeroUnauthorized, XeroAccessDenied
 )
 
 
 OAUTH_EXPIRY_SECONDS = 3600 # Default unless a response reports differently
+DEFAULT_SCOPE = ['openid', 'profile', 'email']
```

Comment:
Should these scopes be constants? There's [a definitive list of available scopes](https://developer.xero.com/documentation/oauth2/scopes) available, which suggests they should.

Reply:
I could add that as `OAUTH2_SCOPES`, which would be a tuple of all available scopes. 

Or do you mean to also define a new constant for each (of the 51) available scopes, like so?
```python
XERO_SCOPE_OPENID = 'openid'
XERO_SCOPE_PROFILE = 'profile'
...
XERO_SCOPE_BANKFEEDS = 'bankfeeds'
```
I'll also change `DEFAULT_SCOPE` to be a tuple.

Finally, should we check the scope passed in to `OAuth2Credentials.__init__()` against the list of available scopes and raise an exception for any scope which is not recognised?

Reply:
I disagree that `DEFAULT_SCOPE` should be a tuple - it's a list of objects of the same type, not an ordered structure of different object types where position is significant. 

51 is... a lot... but having them as constants means that any typos in the scope are caught as linting errors (or, worse case, as symbolic errors client side), rather than failing at runtime when the server rejects or ignores them. 

However, I don't think we need to validate against a known list. Providing a list of known constants is a safety mechanism, but I'm sure Xero will add new scopes in future, in which case it would be nice to be able to support those scopes in the interim between PyXero adding support for those scopes.

Reply:
Sounds good, I will add them all as constants. My reasoning for `DEFAULT_SCOPE` being a tuple was mutability, i.e. to avoid the user accidentally modifying the default scope for all credentials.  I'll leave it as it is now (where `[:]` is used to copy the the default scope on assignment).

Reply:
This has been addressed in the latest commit, please review.